week7/answers.sql

```diff
@@ -1,3 +1,207 @@
+    armor INT,
+    damage INT
+);
+
+CREATE TABLE inventory (
+    inventory_id INT AUTO_INCREMENT PRIMARY KEY,
+    character_id INT,
+    item_id INT,
+    FOREIGN KEY (character_id) REFERENCES characters(character_id),
+    FOREIGN KEY (item_id) REFERENCES items(item_id)
+);
+
+CREATE TABLE equipped (
+    equipped_id INT AUTO_INCREMENT PRIMARY KEY,
+    character_id INT,
+    item_id INT,
+    FOREIGN KEY (character_id) REFERENCES characters(character_id),
+    FOREIGN KEY (item_id) REFERENCES items(item_id)
+);
+
+CREATE VIEW character_items AS
+SELECT DISTINCT inv.character_id, c.name AS character_name, i.name AS item_name, i.armor, i.damage
+FROM inventory inv
+JOIN characters c ON inv.character_id = c.character_id
+JOIN items i ON inv.item_id = i.item_id
+UNION
+SELECT DISTINCT eq.character_id, c.name AS character_name, i.name AS item_name, i.armor, i.damage
+FROM equipped eq
+JOIN characters c ON eq.character_id = c.character_id
+JOIN items i ON eq.item_id = i.item_id;
+
+CREATE VIEW team_items AS
+SELECT tm.team_id, t.name AS team_name, i.name AS item_name, i.armor, i.damage
+FROM team_members tm
+JOIN teams t ON tm.team_id = t.team_id
+JOIN characters c ON tm.character_id = c.character_id
+JOIN inventory inv ON c.character_id = inv.character_id
+JOIN items i ON inv.item_id = i.item_id
+UNION
+SELECT tm.team_id, t.name AS team_name, i.name AS item_name, i.armor, i.damage
+FROM team_members tm
+JOIN teams t ON tm.team_id = t.team_id
+JOIN characters c ON tm.character_id = c.character_id
+JOIN equipped eq ON c.character_id = eq.character_id
+JOIN items i ON eq.item_id = i.item_id;
+
+DELIMITER //
+CREATE FUNCTION armor_total(character_id INT) RETURNS INT
+BEGIN
+    DECLARE total_armor INT;
+    
+    SELECT COALESCE(SUM(cs.armor), 0)
+    INTO total_armor
+    FROM character_stats cs
+    WHERE cs.character_id = character_id;
+    
+    SELECT COALESCE(SUM(i.armor), 0)
+    INTO total_armor
+    FROM equipped eq
+    JOIN items i ON eq.item_id = i.item_id
+    WHERE eq.character_id = character_id;
+    
+    RETURN total_armor;
+END;
+DELIMITER ;
+
+DELIMITER //
+CREATE PROCEDURE attack(id_of_character_being_attacked INT, id_of_equipped_item_used_for_attack INT)
+BEGIN
+    DECLARE total_armor INT;
+    DECLARE total_damage INT;
+    DECLARE char_health INT;
+    
+    SET total_armor = armor_total(id_of_character_being_attacked);
+    
+    SELECT damage INTO total_damage
+    FROM items
+    WHERE item_id = id_of_equipped_item_used_for_attack;
+    
+    SELECT health INTO char_health
+    FROM character_stats
+    WHERE character_id = id_of_character_being_attacked;
+    
+    IF (total_damage - total_armor) > 0 THEN
+        UPDATE character_stats
+        SET health = health - (total_damage - total_armor)
+        WHERE character_id = id_of_character_being_attacked;
+        
+        IF (health - (total_damage - total_armor)) <= 0 THEN
+            DELETE FROM characters WHERE character_id = id_of_character_being_attacked;
```

Comment:
All of these DELETE FROM calls will go away if you simply set your ON UPDATE and ON DELETE behavior for your FOREIGN KEY CONSTRAINTs in you tables.